Fix MSVC build issue from MbedTls issue 7087

ccrugoPhilips opened the pull request:
## Description

Refactored code causing build issues with MSVC toolset v142. This has been tracked under MbedTls issue 7087.



## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [ ] **changelog** provided | not required because: 
- [ ] **framework PR** provided Mbed-TLS/mbedtls-framework# | not required
- [ ] **mbedtls development PR** provided Mbed-TLS/mbedtls# | not required because: 
- [ ] **mbedtls 3.6 PR** provided Mbed-TLS/mbedtls# | not required because: 
- **tests**  provided | not required because: 



## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/TF-PSA-Crypto/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

Help make review efficient:
* Multiple simple commits
  - please structure your PR into a series of small commits, each of which does one thing
* Avoid force-push
  - please do not force-push to update your PR - just add new commit(s)
* See our [Guidelines for Contributors](https://mbed-tls.readthedocs.io/en/latest/reviews/review-for-contributors/) for more details about the review process.

ronald-cron-arm reviewed (changes requested):
Thanks for this contribution. We are not comfortable with the changes in crypto.h. Please see my comment below.

ccrugoPhilips commented on include/psa/crypto.h:
```diff
@@ -99,15 +99,15 @@ psa_status_t psa_crypto_init(void);
+/** Assign an initial value for a key attributes structure.
  */
-static psa_key_attributes_t psa_key_attributes_init(void);
+static void psa_key_attributes_init_internal(psa_key_attributes_t *attributes);
```
Ok, will take a look to your solution and rework this PR. And sure, would be nice to keep crypto_extra.h changes here.

ccrugoPhilips replied:
Moving the include is indeed much simpler to fix build errors from crypto.h and crypto_struct.h; I'm not very familiar with the API so tried to avoid that. Will rollback my changes in those files. 
I see in https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/260 that some non-regression testing for https://github.com/Mbed-TLS/mbedtls/issues/7087, should that be added in this PR?

ccrugoPhilips commented:
A couple of additional remarks:

The build issues reported in https://github.com/Mbed-TLS/mbedtls/issues/7087 are not present anymore with VS2022 (MSVC toolset v143).

Added an extra fix for a warning being treated as error when MBEDTLS_PSA_KEY_STORE_DYNAMIC is disabled.

ronald-cron-arm reviewed (dismissed):
Sorry for the delay.
Two nitpicks but otherwise this looks god to me. Unfortunately, this needs to be rebased though (conflicts with crypto_extra.h).
I've removed the commits related to crypto_extra.h in #260 and added there a change log that should cover both PRs.
Thus if possible on your side, please rebase. Otherwise if this is more convenient for you I can cherry-pick your commits in #260 and resolve the conflicts.
Otherwise @irwir proposed [here](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/260#issuecomment-2912273017) a simpler solution it seems. Would this work for you?

ccrugoPhilips commented:
> Sorry for the delay. Two nitpicks but otherwise this looks god to me. Unfortunately, this needs to be rebased though (conflicts with crypto_extra.h). I've removed the commits related to crypto_extra.h in #260 and added there a change log that should cover both PRs. Thus if possible on your side, please rebase. Otherwise if this is more convenient for you I can cherry-pick your commits in #260 and resolve the conflicts. Otherwise @irwir proposed [here](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/260#issuecomment-2912273017) a simpler solution it seems. Would this work for you?

Hi there, I see that related PRs have already been merged and seems this PR is not necessary anymore; can you please confirm to close this PR. Happy the issue was finally addressed ;)

irwir commented:
> Happy the issue was finally addressed

Addressed, but not fully resolved.
You were given a link to my comment in the merged @ronald-cron-arm's PR; please read.

irwir commented:
> Eventually I do not think we can just remove the declarations

What is the reason?
If no code between forward declaration and function definition references the function, then forward declaration is pointless and safely could be removed.

irwir commented:
> alignment with how things are done for the other xyz_init() functions for the other algorithms in crypto.h

Alignment could have the opposite direction - in my case commenting out all forward declarations fixed the issue (4 lines in crypto.h,  2 lines in crypto_extra.h). But whatever, the current code is far from final.

ronald-cron-arm commented:
Thus I have aligned the changes done to crypto_extra.h with the ones done in https://github.com/Mbed-TLS/mbedtls/pull/10200. The result was https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/258/commits/01284e6c4dbd4a7f42dee8a8fab9847ec8defc2c. Then I rebased to resolve the conflicts in crypto_extra.h due to the work on PAKE key confirmation that was merged as part of another thread of work.

minosgalanakis commented on core/psa_crypto_slot_management.c:
```diff
@@ -660,7 +660,7 @@ psa_status_t psa_reserve_free_key_slot(psa_key_id_t *volatile_key_id,
              * selected_slot but rather unused_persistent_key_slot.  */
             slot_idx = selected_slot - global_data.key_slots;
-            *volatile_key_id = PSA_KEY_ID_VOLATILE_MIN + slot_idx;
+            *volatile_key_id = PSA_KEY_ID_VOLATILE_MIN + (psa_key_id_t) slot_idx;
```
As [discussed](https://github.com/Mbed-TLS/mbedtls/pull/10200#pullrequestreview-2937148566) on the back-port PR was this fix required ?

minosgalanakis commented on include/psa/crypto_extra.h:
```diff
@@ -756,6 +756,234 @@ typedef uint32_t psa_pake_primitive_t;
+    psa_pake_family_t family;
+    uint16_t  bits;
+    psa_algorithm_t hash;
+    uint32_t key_confirmation;
```
This is a note for other reviewers: This API incompatibility is expected and introduced in #200

minosgalanakis reviewed (approved):
Looks good overall, in aligment with https://github.com/Mbed-TLS/mbedtls/pull/10200 with only one missing bugfix on the backport.

davidhorstmann-arm reviewed (approved):
LGTM, thanks!
